dtm0/service.c

```diff
@@ -0,0 +1,625 @@
+	 */
+	struct m0_clink         dop_ha_link;
+	/**
+	 * Link connected to remote process
+	 */
+	struct m0_rpc_link      dop_rlink;
+	/**
+	 * Remote process fid
+	 */
+	struct m0_fid           dop_rproc_fid;
+	/**
+	 * Remote service fid
+	 */
+	struct m0_fid           dop_rserv_fid;
+	/**
+	 * Remote process endpoint
+	 */
+	const char             *dop_rep;
+	/**
+	 * Current dtm0 service dtm0 process to.
+	 */
+	struct m0_reqh_service *dop_dtm0_service;
+	/**
+	 * Connect ast
+	 */
+	struct m0_sm_ast        dop_service_connect_ast;
+	struct m0_clink         dop_service_connect_clink;
+};
+
+M0_TL_DESCR_DEFINE(dopr, "dtm0_process", static, struct dtm0_process, dop_link,
+		   dop_magic, 0x8888888888888888, 0x7777777777777777);
+M0_TL_DEFINE(dopr, static, struct dtm0_process);
+
+/**
+ * typed container_of
+ */
+enum { DTM0_SERVICE_MAGIX = 0x9999999999999999 };
+static const struct m0_bob_type dtm0_service_bob = {
+	.bt_name = "dtm0 service",
+	.bt_magix_offset = M0_MAGIX_OFFSET(struct m0_dtm0_service, dos_magix),
+	.bt_magix = DTM0_SERVICE_MAGIX,
+	.bt_check = NULL
+};
+M0_BOB_DEFINE(static, &dtm0_service_bob, m0_dtm0_service);
+
+static struct m0_dtm0_service *to_dtm(struct m0_reqh_service *service)
+{
+	return bob_of(service, struct m0_dtm0_service, dos_generic,
+		      &dtm0_service_bob);
+}
+
+/**
+ * Service part
+ */
+static int dtm0_service__init(struct m0_dtm0_service *s)
+{
+	m0_dtm0_service_bob_init(s);
+	dopr_tlist_init(&s->dos_processes);
+	m0_dtm0_dtx_domain_init();
+
+	return m0_dtm0_clk_src_init(&s->dos_clk_src, M0_DTM0_CS_PHYS);
+}
+
+static void dtm0_service__fini(struct m0_dtm0_service *s)
+{
+	m0_dtm0_clk_src_fini(&s->dos_clk_src);
+	m0_dtm0_dtx_domain_fini();
+	dopr_tlist_fini(&s->dos_processes);
+	m0_dtm0_service_bob_fini(s);
+}
+
+static struct dtm0_process *dtm0_service_process__lookup(struct m0_reqh_service *reqh_dtm0_svc,
+							 const struct m0_fid *remote_dtm0)
+{
+	struct m0_dtm0_service *dtm0 =
+		container_of(reqh_dtm0_svc, struct m0_dtm0_service, dos_generic);
+	struct dtm0_process *process;
+
+	process = m0_tl_find(dopr, proc, &dtm0->dos_processes,
+			     m0_fid_eq(&proc->dop_rserv_fid, remote_dtm0));
+	M0_ASSERT(process != NULL);
+
+	return process;
+}
+
+
+M0_INTERNAL int m0_dtm0_service_process_connect(struct m0_reqh_service *s,
+						struct m0_fid *remote_srv,
+						const char    *remote_ep,
+						bool async)
+{
+	struct dtm0_process *process;
+	struct m0_rpc_machine *mach =
+		m0_reqh_rpc_mach_tlist_head(&s->rs_reqh->rh_rpc_machines);
+	int rc;
+
+	process = dtm0_service_process__lookup(s, remote_srv);
+	if (process == NULL)
+		return M0_RC(-ENOENT);
+
+	rc = m0_rpc_link_init(&process->dop_rlink, mach, remote_srv,
+			      remote_ep, 10);
+	M0_ASSERT(rc == 0);
+
+	M0_LOG(M0_DEBUG,
+	       " async=%d"
+	       " dtm0="FID_F
+	       " remote_srv="FID_F,
+	       !!async,
+	       FID_P(&s->rs_service_fid),
+	       FID_P(remote_srv));
+	M0_LOG(M0_DEBUG, "rep=%s", remote_ep);
+
+	if (!async)
```

Comment:
rc = !async ? m0_rpc_link_connect_sync(&process->dop_rlink,
					       M0_TIME_NEVER) :
		m0_rpc_link_connect_async(&process->dop_rlink,
					  M0_TIME_NEVER,
					  &process->dop_service_connect_clink);

Reply:
m0_rpc_link_connect_async() is void function.